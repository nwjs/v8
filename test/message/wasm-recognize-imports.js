// Copyright 2023 the V8 project authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// Flags: --experimental-wasm-stringref --allow-natives-syntax
// Flags: --trace-wasm-inlining --liftoff

d8.file.execute('test/mjsunit/wasm/wasm-module-builder.js');

let builder = new WasmModuleBuilder();
let sig_w_w = makeSig([kWasmStringRef], [kWasmStringRef]);
let toLowerCase = builder.addImport("m", "toLowerCase", sig_w_w);

builder.addFunction('call_tolower', sig_w_w).exportFunc().addBody([
  kExprLocalGet, 0,
  kExprCallFunction, toLowerCase,
]);

let module = builder.toModule();

let recognizable = Function.prototype.call.bind(String.prototype.toLowerCase);
let recognizable_imports = { m: { toLowerCase: recognizable } };

let instance1 = new WebAssembly.Instance(module, recognizable_imports);
let call_tolower = instance1.exports.call_tolower;
call_tolower("ABC");
%WasmTierUpFunction(call_tolower);
call_tolower("ABC");

// Creating a second instance with identical imports should not cause
// recompilation.
console.log("Second instance.");
let instance2 = new WebAssembly.Instance(module, recognizable_imports);
let call_tolower2 = instance2.exports.call_tolower;
call_tolower2("DEF");
console.log("Still optimized: " + %IsTurboFanFunction(call_tolower2));

// Creating a third instance with different imports must not reuse the
// existing optimized code.
console.log("Third instance.");
let other_imports = { m: { toLowerCase: () => "foo" } };
let instance3 = new WebAssembly.Instance(module, other_imports);
let call_tolower3 = instance3.exports.call_tolower;
call_tolower3("GHI");
console.log("Still optimized: " + %IsTurboFanFunction(call_tolower3));
