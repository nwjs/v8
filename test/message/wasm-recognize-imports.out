[function 1: import 0 is well-known built-in String.toLowerCase]
Second instance.
Still optimized: true
Third instance.
Still optimized: false
